### ISSUE TITLE
fields not updating after re-calc

### ISSUE THREAD

**Issue by syfgk** [mortii/anki-morphs#220]
### Describe the bug

[//]: # (Please provide a clear and concise description of what the bug is.)
(For context, I'm using AnkiMorphs with the Chinese SpaCy models.)

Here's the bug: after running re-calc, some AnkiMorphs fields are not updating.

Below is one example. Notice that `am-unknowns` and `am-unknowns-count` are populated. However, I'd only expect this for new cards. This card, though, is one I've reviewed four times already. Because it's seen, I'd expect `am-unknowns` to be null and `am-unknowns-count` to be zero. Am I misunderstanding how re-calc works?

<img width="461" alt="image" src="https://github.com/mortii/anki-morphs/assets/164821375/aa8b3738-4e00-4479-af62-55a33bbd4f27">

<br>

<img width="336" alt="image" src="https://github.com/mortii/anki-morphs/assets/164821375/7545273b-18d5-4e3d-8976-9e41bce70b59">

I should note that this is not affecting all seen cards. There are other (seen) cards that are being re-calc'd as I'd expect (i.e., `am-unknowns` and `am-unknowns-count` are empty and 0, respectively).

### My setup

 - Operating System: MacOS
 - Anki Version:  Version ⁨23.12.1 (1a1d4d54)⁩; Python 3.9.15 Qt 6.5.3 PyQt 6.5.3
 - AnkiMorphs Version: v2.2.2

**Comment by mortii** [comment]
> Here's the bug: after running re-calc, some AnkiMorphs fields are **_not_** updating.

This seems to directly contradict this:

> Notice that am-unknowns and am-unknowns-count **_are populated_**. However, I'd only expect this for new cards.

I'll assume you meant that the extra fields are updating when the shouldn't.

The only extra field that should update on non-new cards is the `am-highlighted` field. In this context "new" means cards not in the new-queue, it might be possible to move cards to the new-queue if you reset them, and you might see the still see the review count if you don't select `Reset repetition and lapse count`, but I haven't checked.

You could double check if the fields are actually updating by erasing all the data in the extra fields and then running recalc again.

**Comment by syfgk** [comment]
Can we walk through a hypothetical example to confirm my understanding of how re-calc works?

Let's say, for instance, I have one new card which has 1 unseen morph. Since this is a card I have not reviewed yet, I would expect that running re-calc would result in the following:
- `am-unknowns` = whatever the 1 unseen morph is
- `am-unknowns-count` = 1

Now, _after_ I review the card, if I run re-calc again, my understanding is that I would now see the following:
- `am-unknowns` = blank
- `am-unknowns-count` = 0

Is my understanding correct?

**Comment by mortii** [comment]
> Now, _after_ I review the card, if I run re-calc again, my understanding is that I would now see the following:
> 
>     * `am-unknowns` = blank
> 
>     * `am-unknowns-count` = 0
> 
> 
> Is my understanding correct?

No, they will keep having the same values as when the card was new, since they don't get updated, in this case:
- `am-unknowns` = whatever
- `am-unknowns-count` = 1

This essentially freezes the card in the same state as when you first review it, that way you can have the `am-unknowns` field visible on your card and not lose it the next time you review. 

So for this card:
![image](https://github.com/mortii/anki-morphs/assets/15674619/a175684e-16e5-4aaa-a9fa-1132e0fc3c25)

I would still see the word on the top, even though it's not technically "unknown" to me anymore. This makes sure that useful context is not lost on future reviews.

**Comment by syfgk** [comment]
Ah, ok. So, follow-up question: why do some cards have:
- am-unknowns = blank
- am-unknowns-count = 0

**Comment by mortii** [comment]
If you have two cards in this order:

|         | card queue status | unknown morphs | unknown morphs count |
|---------|-------------------|----------------|----------------------|
| Card 01 | new               | whatever       | 1                    |
| Card 02 | new               | whatever       | 1                    |


and you learn the first card, then the morphs on that first card will no longer be unknown on any new cards:

|         | card queue status | unknown morphs | unknown morphs count |
|---------|-------------------|----------------|----------------------|
| Card 01 | review            | whatever       | 1                    |
| Card 02 | new               |                | 0                    |


does that make sense?

**Comment by syfgk** [comment]
It's making more sense! In this case we're assuming the "unknown morphs" are the same across both cards, right?

**Comment by mortii** [comment]
> In this case we're assuming the "unknown morphs" are the same across both cards, right?

Sorry, it wasn't a great example, but yes, both cards have the same unknown morph "whatever". Substitute it with "airplaine" or "人" or any other morph.

Btw, I'm not saying this current approach is good, in fact, the `am-unknowns-count` field should probably update for all cards regardless if they are new or not, that way the sorting by that field would be more useful. Do you have any suggestions/preferences for a better approach?

**Comment by syfgk** [comment]
Ah! I see, ok things are making sense now. So let me ask this:

Let's say I have these cards, both are new:

|        | card queue status | unknown morphs | unknown morphs count |
|--------|-------------------|----------------|----------------------|
| Card 1 | new               | A, B           | 2                    |
| Card 2 | new               | A              | 1                    |

Now let's say I review the first card. My understanding is that this is what I'd now see:

|        | card queue status | unknown morphs | unknown morphs count |
|--------|-------------------|----------------|----------------------|
| Card 1 | review               | A, B           | 2                    |
| Card 2 | new               |                | 0                    |

Is this correct?

**Comment by mortii** [comment]
Yep, both of those scenarios are correct, although the first one should realistically only happen during some custom review circumstances or if they are in different decks.

**Comment by syfgk** [comment]
Nice! Here's my suggestion: have _**four**_ extra fields instead of two

- `unknown-morphs`: shows all the unknown/unseen morphs on a new card. Once a card has been reviewed/seen and re-calc is run, this field goes blank

- `unknown-morphs-count`: displays the count of `unknown-morphs`. Once a card has been reviewed/seen and re-calc is run, this field also goes blank

- _**(new)**_ `all-morphs`: this shows all the morphs on the card, regardless of card state. So even after the user has reviewed/seen the card, this field stays the same

- _**(new)**_ `all-morphs-count`: this is a count of `all-morphs`; same logic applies. After I’ve reviewed/seen the card, this field also stays the same

Here’s an example:

Table 1 (card is new)

|        | card queue status | unknown morphs | unknown morphs count | all morphs | all morphs count |
|--------|-------------------|----------------|----------------------|------------|------------------|
| Card 1 | new               | A,B,C,D        | 4                    | A,B,C,D    | 4                |

Table 2 (card has been reviewed)

|        | card queue status | unknown morphs | unknown morphs count | all morphs | all morphs count |
|--------|-------------------|----------------|----------------------|------------|------------------|
| Card 1 | review               |         | 0                    | A,B,C,D    | 4                |

What do you think?

**Comment by mortii** [comment]
That is very logical, and I don't mind adding new extra fields (I coincidentally reached the same conclusion about adding an "all morphs count" extra field in #195)

I don't really see how prevent the loss of the unknown morph after the first review though, like I mentioned in my previous comment https://github.com/mortii/anki-morphs/issues/220#issuecomment-2053691213, that context can be pretty important later. Any ideas?

**Comment by syfgk** [comment]
Is the image in that comment both sides of your card, or just one side? I want to make sure I'm oriented correctly and know what's on the front and bank of your card.

**Comment by mortii** [comment]
Here is another example card:

![Screenshot from 2024-04-14 15-14-34](https://github.com/mortii/anki-morphs/assets/15674619/15e8790f-dbb6-47de-9cce-c0189d46261d)
![Screenshot from 2024-04-14 15-15-00](https://github.com/mortii/anki-morphs/assets/15674619/d08c39cc-3d27-4d35-aada-75d50c3eac28)

the template looks something like this (simplified):
![image](https://github.com/mortii/anki-morphs/assets/15674619/4f6078f5-aca9-4ab6-b853-2829234c6646)

where the `am-unknowns` extra field is show on the front and back.

**Comment by syfgk** [comment]
Thanks for sharing! What's the value of showing `am-unknowns` on the front instead of just having `am-highlighted`?
